### PR TITLE
[reland] Add forward mode AD to out-place foreach functions

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes https://github.com/pytorch/pytorch/pull/102409#issuecomment-1615254393

This introduces `make_saved_variable_list_foreach_output` which creates a `vector<SavedVariable>` with `is_output=true` unlike the other `make_saved_variable_list` implementations that are used for functions inputs apparently.

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -918,39 +918,62 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
...
+        # note(crcrpar): without this, some unary functions fail, unlike inplace and/or complex.
+        if dtype == torch.float64 and op.name in (
+            "_foreach_acos", "_foreach_asin", "_foreach_log10", "_foreach_log1p", "_foreach_log2",
+            "_foreach_log", "_foreach_pow", "_foreach_sqrt",
+        ):
+            value_range = {"low": 0.5, "high": 1.0}
+        else:
+            value_range = {}
+        for sample in op.sample_inputs(
+            device, dtype, requires_grad=False, num_input_tensors=[5], same_size=True, **value_range,
```
Comment (crcrpar) [inline test/test_foreach.py:934]: `requires_grad=False` is the difference from the original pull request and what I'm unhappy with at the moment.

File: torch/autograd/gradcheck.py
Comment (crcrpar) [inline torch/autograd/gradcheck.py:1]: forward-mode AD checks in test_foreach.py look green with `requires_grad=False`. I do believe this is temporary and I can manage to remove this argument.

**Comment by crcrpar** [comment]
hmm, IIUC saving result tensors to grad_fn is expected as per https://github.com/pytorch/pytorch/issues/95431#issuecomment-1452451618 but this seems to explain the cuda memory leak in forward grad tests.

How should I handle this?

**Comment by albanD** [comment]
I'll let @soulitzer handle the first pass on this.

**Comment by soulitzer** [comment]
Hmm sounds like a preexisting foreach backward AD issue? If that is the case we should figure out why that wasn't being tested before, and do the fix/add tests in a separate PR from this. Then we can land the original forward AD out-of-place one as-is.

**Comment by crcrpar** [comment]
> a preexisting foreach backward AD issue?

I think so. `tools/autograd/gen_variable_type` seems to be responsible for the mem leak.

> why that wasn't being tested before

I had not noticed this, and the tests added when the backward AD was added didn't cover the case where `grad_fn->result_` is set but not used.

**Comment by crcrpar** [comment]
close in favor of #105504

**Pull request closed without merging**